fabtests/hpcs/include/test/user.h

```diff
@@ -0,0 +1,492 @@
+
+	/*
+	 * Number of fi_contexts to allocate for per-transfer state.
+	 * Also affects number of counter events expected per transfer.
+	 */
+	size_t tx_context_count;
+	size_t rx_context_count;
+
+	uint64_t mr_rx_flags;
+};
+
+typedef struct test_config (*test_config)(
+	const struct test_arguments *arguments
+);
+
+enum op_state {
+	OP_DONE = 0,
+	OP_PENDING
+};
+
+struct context_info {
+	union {
+		struct fi_context2		fi_context;
+		struct fi_triggered_context2	fi_trig_context;
+	};
+	union {
+		struct fi_msg_tagged	tagged;
+		struct fi_msg_rma	rma_msg;
+		struct fi_msg_atomic	atomic_msg;
+	};
+	union {
+		struct iovec		iov;
+		struct fi_ioc		ioc;
+	};
+	union {
+		struct fi_rma_iov	rma_remote_iov;
+		struct fi_rma_ioc	rma_remote_ioc;
+	};
+	struct fi_deferred_work def_work;
+	union {
+		struct fi_op_rma	rma_op;
+		struct fi_op_atomic	atomic_op;
+		struct fi_op_tagged	tagged_op;
+	};
+	struct op_context	*op_context;
+};
+
+struct op_context {
+	struct context_info	*ctxinfo;
+	enum op_state		state;
+	uint8_t			*buf;
+	uint64_t		core_context;
+	uint64_t		test_context;
+	struct fid_mr		*tx_mr;
+	struct fid_cntr		*tx_cntr;
+	struct fid_domain	*domain;
+	uint64_t		test_state; /* reserved for test internal accounting */
+};
+
+/*
+ * DATA OBJECT SETUP
+ *
+ * Setup any required data objects, including filling the buffers with initial
+ * data, creating memory regions, and associating buffers with the memory
+ * regions.  The memory backing these data objects is allocated and freed by the
+ * caller.
+ *
+ * If the test configuration specifies that a memory region should be used for
+ * RX, then in addition to test_rx_init_buffer, a corresponding
+ * test_rx_create_mr will also be called.  test_tx_create_mr will be called if
+ * the provider sets the FI_MR_LOCAL mode bit.
+ *
+ * arguments: the parsed version of the arguments.
+ *
+ * buffer: the buffer to initialize (and associate with a memory region).
+ *
+ * key: the key with which to create the memory region (decided automatically by
+ * the caller).
+ *
+ * buffer: the buffer to initialize and possibly associate with a memory region
+ * (allocated automatically by the caller).
+ */
+
+typedef void (test_tx_init_buffer)(const struct test_arguments *arguments,
+		uint8_t *buffer, size_t len);
+
+test_tx_init_buffer test_generic_tx_init_buffer;
+
+typedef void (test_rx_init_buffer)(const struct test_arguments *arguments,
+		uint8_t *buffer, size_t len);
+
+test_rx_init_buffer test_generic_rx_init_buffer;
+
+typedef struct fid_mr *(test_tx_create_mr)(const struct test_arguments *arguments,
+		struct fid_domain *domain, const uint64_t key,
+		uint8_t *buffer, size_t len, uint64_t access, uint64_t flags);
+
+test_tx_create_mr test_generic_tx_create_mr;
+
+typedef struct fid_mr *(test_rx_create_mr)(const struct test_arguments *arguments,
+		struct fid_domain *domain, const uint64_t key,
+		uint8_t *buffer, size_t len, uint64_t access, uint64_t flags);
+
+test_rx_create_mr test_generic_rx_create_mr;
```

Comment:
These don't have to be typedefed.  We could have the canonical function prototypes be in the definition of  "struct test_api", and then declare an instance of that structure with the generic function implementations, rather than declaring each generic function individually at the top level.

We could also split struct_api into two sub-structures for tx and rx, so we don't have to repeat the same function prototypes twice.